packages/core/echo/echo-schema/src/automerge/automerge-db.ts

```diff
@@ -218,36 +231,125 @@ export class AutomergeDb {
+    handle.change((newDoc: SpaceDoc) => {
+      newDoc.access = { spaceKey: this.spaceKey.toHex() };
+    });
+  }
+
+  private _linkObjectDocument(object: AutomergeObject, handle: DocHandle<SpaceDoc>) {
+    this._objectDocumentHandles.set(object.id, handle);
+    this._spaceRootDocHandle.change((newDoc: SpaceDoc) => {
+      if (newDoc.links) {
+        newDoc.links[object.id] = handle.url;
+      } else {
+        newDoc.links = { [object.id]: handle.url };
+      }
+    });
+  }
+
+  private _loadLinkedObjects(links: SpaceDoc['links']) {
+    if (!links) {
+      return;
+    }
+    for (const [objectId, automergeUrl] of Object.entries(links)) {
+      if (this._objectDocumentHandles.has(objectId)) {
+        return;
+      }
+      const handle = this.automerge.repo.find<SpaceDoc>(automergeUrl as DocumentId);
+      log.debug('document loading triggered', { objectId, automergeUrl });
+      this._objectDocumentHandles.set(objectId, handle);
+      this._createObjectOnDocumentLoad(objectId, handle);
+    }
+  }
+
+  private _onDocumentUpdate(event: DocHandleChangePayload<SpaceDoc>) {
+    const { updatedObjects, linkedDocuments } = processDocumentUpdate(event);
+    this._loadLinkedObjects(linkedDocuments);
+    this._createInlineObjects(
+      event.handle,
+      updatedObjects.filter((id) => !this._objects.has(id)),
+    );
+    this._emitUpdateEvent(updatedObjects);
+  }
+
+  private _onDispose() {
+    this._spaceRootDocHandle?.off('change');
+    for (const docHandle of this._objectDocumentHandles.values()) {
+      docHandle.off('change');
+    }
+  }
+
   /**
    * Loads all objects on open and handles objects that are being created not by this client.
    */
-  private _createObjects(objectIds: string[]) {
-    invariant(this._docHandle);
+  private _createInlineObjects(docHandle: DocHandle<SpaceDoc>, objectIds: string[]) {
     for (const id of objectIds) {
       invariant(!this._objects.has(id));
-      const obj = new AutomergeObject();
-      obj[base]._core.id = id;
-      this._objects.set(obj.id, obj);
-      (obj[base] as AutomergeObject)._bind({
-        db: this,
-        docHandle: this._docHandle,
-        path: ['objects', obj.id],
-        assignFromLocalState: false,
-      });
+      this._createObjectInDocument(id, docHandle);
     }
   }
-}
 
-const getUpdatedObjects = (event: DocHandleChangePayload<DocStructure>): string[] => {
-  const updatedObjects = event.patches
-    .map(({ path }: { path: string[] }) => {
-      if (path.length >= 2 && path[0] === 'objects') {
-        return path[1];
-      }
-      return undefined;
-    })
-    .filter(Boolean);
+  private _createObjectOnDocumentLoad(objectId: string, handle: DocHandle<SpaceDoc>) {
+    handle
+      .doc(['ready'])
+      .then(() => {
+        if (this._ctx?.disposed ?? true) {
+          log.warn('document loaded after database was closed');
+          return;
+        }
+        handle.on('change', this._onDocumentUpdate.bind(this));
```

Comment:
clear those subscriptions when db closes: `this._ctx.onDispose(() => ....)`

Reply:
this is handled in `_onDispose`